Audit3 channel transfer extrinsics blocked

dobertRowneySr opened the pull request:
Addresses #3428 (last comment by Bedeho)

#### Changes
- [x] added guard for each extrinsic that should be blocked during transfer
- [x] remove guard from metaprotocols extrinsics
- [x] added test for each extrinsic during scenario
- [x] rebased on lates `carthage` & updated polkadot-js packages

L3pereira committed:
introduce_channel_state_bloat_bond_#3511

requested changes

Carthage T: Trait fix

L3pereira committed:
Introduce video state bloat bond #3394

Introduce video state bloat bond #3394

dobertRowneySr committed:
test fixes

dobertRowneySr committed:
fix tests

dobertRowneySr committed:
fix tests

dobertRowneySr committed:
add revenue_split_rate to token data

dobertRowneySr committed:
add revenue_split_rate to token data

dobertRowneySr committed:
rebase

L3pereira committed:
remove_council_blog_pallet_and_proposals_#3662

remove_blog_proposals_#3662

remove_council_blog_pallet_and_proposals_#3662

remove_council_blog_typescript_pallet_and_proposals_#3662

remove_council_blog_typescript_pallet_and_proposals_#3662

requested changes + cargo fmt in types.rs

requests + types + metadata

chain metadata

dobertRowneySr committed:
block/unblock extrinsics during channel transfer

dobertRowneySr committed:
add test for blocked extr.s during channel transfers

dobertRowneySr committed:
remove categories from content

dobertRowneySr committed:
fix runtime

dobertRowneySr committed:
fix chainspec config

dobertRowneySr committed:
generate types library and metadata

L3pereira committed:
video state bloat bond Carthage T: Trait fix

dobertRowneySr committed:
aremove channel caterogy id from Genesis Config

dobertRowneySr committed:
fix clippy

dobertRowneySr committed:
generate types

dobertRowneySr committed:
generate types

dobertRowneySr committed:
clippy fix

dobertRowneySr committed:
updated packages

dobertRowneySr committed:
set next_sale_id starting value to 0

dobertRowneySr committed:
set next_sale_id starting value to 0

dobertRowneySr committed:
add test with failing destination member

dobertRowneySr committed:
rename allocation_amount -> revenue_amount
So the terminology is:
- allocation amount = what creator allocates for a revenue split
- revenue amount = what is actually provided to shareholders

dobertRowneySr committed:
disallow issuing a token with 0 revenue split rate

dobertRowneySr committed:
add test for lefover revenue split allocation amount

dobertRowneySr committed:
generate metadata & types

dobertRowneySr committed:
yarn format

mnaamani committed:
at least one nomination if initial nominators setup at genesis

L3pereira committed:
types generation and channel duplication fix

types generation

types generation and channel duplication fix

L3pereira committed:
types generation

dobertRowneySr committed:
swap names allocation <-> revenue

dobertRowneySr committed:
swap revenue <-> allocation in tests & traits

dobertRowneySr committed:
update chain meta data

dobertRowneySr committed:
send fund to controller account after rev. split

dobertRowneySr committed:
add test for leftover funds

Lezek123 commented on runtime-modules/content/src/tests/creator_tokens/finalize_creator_token_sale.rs:
```diff
@@ -213,3 +214,31 @@ fn successful_finalize_curator_channel_creator_token_sale_by_lead() {
+            ContentTreasury::<Test>::account_for_channel(ChannelId::one()),
+            DEFAULT_PAYOUT_EARNED
+                .saturating_add(<Test as balances::Config>::ExistentialDeposit::get().into()),
+        );
```
This is not needed for sales (only for revenue splits)

Lezek123 commented on runtime-modules/content/src/tests/nft/accept_incoming_offer.rs:
```diff
@@ -263,3 +263,23 @@ fn accept_incoming_offer_insufficient_balance() {
+
+        assert_noop!(
+            Content::accept_incoming_offer(
+                Origin::signed(DEFAULT_MEMBER_ACCOUNT_ID),
```
```suggestion
                Origin::signed(SECOND_MEMBER_ACCOUNT_ID),
```

Lezek123 reviewed (changes requested):
Looks good, but some tests/features are still missing in order for channel transfers blocking feature to be complete

**Missing test cases for:**
- make_open_auction_bid
- make_english_auction_bid
- cancel_open_auction_bid
- settle_english_auction
- sell_nft

<s>
**Missing feature: `channel_transfer_id`**
See the comment about `channel_transfer_id` in https://github.com/Joystream/joystream/issues/3428, it is quite an important feature, so I think it would be nice to add it in the scope of this PR
</s>

_Nevermind, introduced in https://github.com/Joystream/joystream/pull/3972_

**Missing feature: transfers not allowed during active creator token sale/revenue split:**
See https://github.com/Joystream/joystream/issues/3428#issuecomment-1146195008 (under **_Creator tokens_**)

dobertRowneySr committed:
fix naming inconsistencies

dobertRowneySr committed:
fix naming inconsistency 2

dobertRowneySr committed:
cargo format

mnaamani committed:
Update node/src/chain_spec/mod.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

dobertRowneySr committed:
open auction bid fails during transfer

dobertRowneySr committed:
transfer fails during english auction

dobertRowneySr committed:
settle english auction fails during transfer

dobertRowneySr committed:
settle eng auction fails during transfer

dobertRowneySr committed:
add tests for transfer blocked during sales revenue split

dobertRowneySr committed:
add tests for transfer blocked during sales revenue split

dobertRowneySr committed:
token revenue splits and sale block transfers

dobertRowneySr committed:
fix test errors during tests

dobertRowneySr committed:
clippy fix

dobertRowneySr committed:
generate types and metadata

dobertRowneySr committed:
clippy warning

dobertRowneySr committed:
generate types & metadata

dobertRowneySr committed:
Merge branch 'carthage' into audit3__channel_transfer_extrinsics_blocked

dobertRowneySr committed:
generate types and metadata

dobertRowneySr committed:
Merge branch 'carthage' into audit3__channel_transfer_extrinsics_blocked

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -704,6 +704,24 @@ impl<T: Config>
     > for Module<T>
 {
+    /// Establish weather revenue split is ongoing
+    /// Postconditions: true if token @ token_id has an ongoing revenue split, fales otherwise
```
```suggestion
    /// Postconditions: true if token @ token_id has an unfinalized revenue split, false otherwise
```

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -704,6 +704,24 @@ impl<T: Config>
         TransfersWithVestingOf<T>,
     > for Module<T>
 {
+    /// Establish weather revenue split is ongoing
```
```suggestion
    /// Establish whether there's an unfinalized revenue split
```

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -704,6 +704,24 @@ impl<T: Config>
+    }
+
+    /// Establish weather revenue sale is ongoing
+    /// Postconditions: true if token @ token_id has an ongoing sale, fales otherwise
```
```suggestion
    /// Establish whether there is an unfinalized token sale
    /// Postconditions: true if token @ token_id has an unfinalized sale, false otherwise
```

Lezek123 commented on runtime-modules/project-token/src/traits.rs:
```diff
@@ -73,4 +73,11 @@ pub trait PalletToken<
     /// Finalize creator token sale and recover unsold tokens
     fn finalize_token_sale(token_id: TokenId) -> Result<JoyBalance, DispatchError>;
+
+    /// Establish weather the token has a revenue split pending
```
```suggestion
    /// Establish whether the token has an unfinalized revenue split
```

Lezek123 commented on runtime-modules/project-token/src/traits.rs:
```diff
@@ -73,4 +73,11 @@ pub trait PalletToken<
+    /// Establish weather the token has a revenue split pending
+    fn is_revenue_split_inactive(token_id: TokenId) -> bool;
+
+    /// Establish weather the token has a sale pending
```
```suggestion
    /// Establish whether the token has an unfinalized sale
```

Lezek123 commented on runtime-modules/content/src/lib.rs:
```diff
@@ -2392,6 +2407,18 @@ decl_module! {
                 Self::validate_member_set(&params.transfer_params.new_collaborators.keys().cloned().collect())?;
             }
 
+            if let Ok(token_id) = channel.ensure_creator_token_issued::<T>() {
+                ensure!(
+                    T::ProjectToken::is_revenue_split_inactive(token_id),
+                    Error::<T>::ChannelTransfersBlockedDuringRevenueSplits
+                );
+
+                ensure!(
+                    T::ProjectToken::is_sale_unscheduled(token_id),
+                    Error::<T>::ChannelTransfersBlockedDuringTokenSales,
+                );
+            }
```
Missing test coverage for those cases (ideally testing upcoming / ongoing / "finished but unfinalized" case for sales and the same for revenue splits)

Lezek123 commented on runtime-modules/content/src/tests/metaprotocol.rs:
```diff
@@ -363,7 +363,7 @@ fn unsuccessful_nft_owner_with_invalid_acount() {
 
 #[test]
-fn unsuccessful_nft_rewark_during_transfer() {
+fn successful_nft_rewark_during_transfer() {
```
```suggestion
fn successful_nft_remark_during_transfer() {
```

Lezek123 commented on runtime-modules/content/src/tests/metaprotocol.rs:
```diff
@@ -395,19 +392,16 @@ fn unsuccessful_channel_owner_remark_during_transfer() {
 
 #[test]
-fn unsuccessful_channel_agent_rewark_during_transfer() {
+fn successful_channel_agent_rewark_during_transfer() {
```
```suggestion
fn successful_channel_agent_remark_during_transfer() {
```

Lezek123 commented on runtime-modules/content/src/tests/creator_tokens/finalize_creator_token_sale.rs:
```diff
@@ -213,3 +214,26 @@ fn successful_finalize_curator_channel_creator_token_sale_by_lead() {
 }
+
+#[test]
+fn unsuccessful_finalize_token_sale_during_channel_transfer() {
```
Now we have `unsuccessful_finalize_token_sale_during_channel_transfer`, but we don't have `unsuccessful_finalize_revenue_split_during_channel_transfer`.

Provided that it should be impossible to init channel transfer if there's an unfinalized token sale / revenue split (and it's also not possible to initialize any of those), I think it's safe to skip both those cases.

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/traits.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/traits.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/content/src/tests/metaprotocol.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/content/src/tests/metaprotocol.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

dobertRowneySr committed:
add test cases for upcoming / ongoing / unfinalized sale

dobertRowneySr committed:
add transfer tests for upcoming / ongoing / unfinalized split

dobertRowneySr committed:
remove test

since purchase_tokens_on_sale helper function basically auto finalizes
the sale

dobertRowneySr committed:
Merge branch 'carthage' into audit3__channel_transfer_extrinsics_blocked

dobertRowneySr committed:
generate metadata

Lezek123 commented on runtime-modules/content/src/tests/nft/cancel_open_auction_bid.rs:
```diff
@@ -298,3 +295,25 @@ fn cancel_open_auction_bid_ok_for_expired_auction() {
 }
+
+#[test]
+fn cancel_open_auction_fails_during_transfer() {
```
```suggestion
fn cancel_open_auction_bid_fails_during_transfer() {
```

Lezek123 reviewed (approved):
LGTM

Commit:
Update runtime-modules/content/src/tests/nft/cancel_open_auction_bid.rs
